src/main/java/bio/terra/app/controller/DatasetsApiController.java

```diff
@@ -510,4 +493,25 @@ private void validateAzureIngestParams(IngestRequestModel ingestRequest) {
       throw new InvalidJobParameterException("Invalid ingest parameters detected", errors);
     }
   }
+
+  private void verifyDatasetAuthorization(
```

Comment:
I have some concerns about this approach. Namely, users can now distinguish between a dataset that exists but they have no access to, and one that does not exist at all. This potentially exposes privileged information. In my opinion, I would also add in that if a user has no access at all to the underlying dataset, that should also return a 404, not a 403. That way, you avoid releasing that information.

Reply:
I think whether a user should know if something exists if they don't have access to it is the main question here. I'm fine with either approach, but the current behavior has confused some of our users. Maybe we just need to make that clearer in the UI/API docs?

Also right now the API is inconsistent, so if a user goes to get a resource that does not exist they will get a 403 error. But if they try to get/update/delete policies on that resource they will get a 404 error. This is the same behavior in Sam. Also as you pointed out, if a user has access to a snapshot but not the source dataset, they can still see all of the source dataset's details through the snapshot details endpoint.

Reply:
Hmm... Sam should also probably change this. I think I maybe filed something about this on the old CI Board.

My general rule is that:
404. Doesn't exist or a user has no permissions to see it
403. A user has insufficient access to perform the action they are trying to perform, but does have some access

Does that make sense?

Reply:
That makes sense! I thought that was the reasoning behind the current behavior (except now I'm confused because I realized the policy endpoints don't follow that pattern).

Reply:
Yeah, there was a conversation around this point and the consensus ended up being (and I agree) if you're accessing via ID like we do in these cases, differentiating 403s and 404s is reasonable from a UX standpoint.  This is especially true since the ID is meaningless.  I'd be nervous if the APIs were something like getDatasetByName since then you could absolutely hacking to find if "organization X" had created a dataset but since these are system generated UUIDs, I think that it's safe.

Reply:
And I'll add the UX is way better this way

Reply:
I talked to Saman in the parking lot - TLDR I think it is okay for this case (although even with UUIDs, since I believe UUIDs are used in the urls for TDR UI, knowing existence would allow someone to construct a URL and say "hey a team member shared this link with me, but I seem to not have access, could you..."), but I don't think Sam solely uses UUIDs for resource IDs (even if it does for TDR resources). So I am going to follow up with these concerns with @gpolumbo-broad, but for now I think this PR can move forwards.